Add `valet link` command

TangRufus opened the pull request:
#### Example

```sh-session
$ trellis-cli valet --help
Usage: trellis valet <subcommand> [<args>]

Subcommands:
    link    Link the local_path directories to Valet

$ trellis-cli valet link --help
Usage: trellis valet link [options] [ENVIRONMENT=development]

Link the local_path directories to Valet

See https://laravel.com/docs/master/valet#the-link-command for more information.

Options:
  -h, --help  show this help

$ trellis-cli valet link
Linking environment development...

Linking site typist.tech...
Running command => valet link --secure typist
A [typist] symbolic link has been created in [/Users/me/.config/valet/Sites/typist].
Restarting nginx...
The [typist.test] site has been secured with a fresh TLS certificate.
✓ Site typist.tech linked

Linking site third.second.com...
Running command => valet link --secure third.second
A [third.second] symbolic link has been created in [/Users/me/.config/valet/Sites/third.second].
Restarting nginx...
The [third.second.test] site has been secured with a fresh TLS certificate.
✓ Site third.second.com linked

Linking site insecure.com...
Running command => valet link insecure
A [insecure] symbolic link has been created in [/Users/me/.config/valet/Sites/insecure].
✓ Site insecure.com linked

Linking site no-enabled.com...
Running command => valet link no-enabled
A [no-enabled] symbolic link has been created in [/Users/me/.config/valet/Sites/no-enabled].
✓ Site no-enabled.com linked

Linking site no-ssl.com...
Running command => valet link no-ssl
A [no-ssl] symbolic link has been created in [/Users/me/.config/valet/Sites/no-ssl].
✓ Site no-ssl.com linked
```

The example `wordpress_sites.yml`:

```yaml
wordpress_sites:
  typist.tech:
    site_hosts:
      - canonical: typist.test
    local_path: ../bedrock
    ssl:
      enabled: true

  third.second.com:
    site_hosts:
      - canonical: third.second.test
    local_path: ../bedrock-third-second
    ssl:
      enabled: true

  insecure.com:
    site_hosts:
      - canonical: insecure.test
    local_path: ../bedrock-insecure
    ssl:
      enabled: false

  no-enabled.com:
    site_hosts:
      - canonical: no-enabled.test
    local_path: ../bedrock-no-enabled
    ssl:
      provider: self-signed

  no-ssl.com:
    site_hosts:
      - canonical: no-ssl.test
    local_path: ../bedrock-no-ssl
```

The example environment:

```sh-session
$ sw_vers
ProductName:	Mac OS X
ProductVersion:	10.14.6
BuildVersion:	18G84

$ php -v
PHP 7.3.7 (cli) (built: Jul  5 2019 12:44:05) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.3.7, Copyright (c) 1998-2018 Zend Technologies
    with Zend OPcache v7.3.7, Copyright (c) 1999-2018, by Zend Technologies

$ composer --version
Composer version 1.8.6 2019-06-11 15:03:05

$ composer global show laravel/valet
Changed current directory to /Users/me/.composer
name     : laravel/valet
descrip. : A more enjoyable local development experience for Mac.
keywords : laravel, wwdhhd, zonda
versions : * v2.3.3
type     : library
license  : MIT License (MIT) (OSI approved) https://spdx.org/licenses/MIT.html#licenseText
source   : [git] https://github.com/laravel/valet.git caf52d66e8eec828cddb57ce0815ab99d8770fd5
dist     : [zip] https://api.github.com/repos/laravel/valet/zipball/caf52d66e8eec828cddb57ce0815ab99d8770fd5 caf52d66e8eec828cddb57ce0815ab99d8770fd5

$ valet --version
Laravel Valet 2.3.3
```

#### Notes

`func TestValetLinkRun(t *testing.T)` is not testing what it should be testing.
I can't get `mockExecCommand` working because `go` complaints when changing working directory.

Linking SSL enabled sites takes time because valet restarts Nginx for each site. `valet link` doesn't provide an option to skip it.

Never tested on [valet-linux](https://github.com/cpriego/valet-linux)

Only the first `canonical` of each site is linked.

Assuming `canonical` TLD is the same as `$ valet tld`.

---

Related #20

swalkinshaw commented on cmd/valet_link.go:
```diff
@@ -0,0 +1,90 @@
+
+	switch len(args) {
+	case 0:
+		environment = "development"
```
Most commands don't have a default env but I guess this makes sense since Valet should be in development only 👍

swalkinshaw commented on cmd/valet_link.go:
```diff
@@ -0,0 +1,90 @@
+
+		isSiteSslEnabled := site.Ssl["enabled"] == true
+
+		valetLink := execCommand("valet", "link", app)
```
Could do this as well:

```go
valetArgs := []string{"link"}

if isSiteSslEnabled {
  valetArgs = append(valetArgs, "--secure")
}

valetLink := execCommand("valet", ...valetArgs, app)
```

TangRufus replied:
`valetLink := execCommand("valet", ...valetArgs, app)` gives syntax error

Have to do:
```go
valetArgs = append(valetArgs, app)
valetLink := execCommand("valet", valetArgs...)
```

swalkinshaw replied:
🤦‍♂ im still new to Go. `execCommand("valet", valetArgs..., app)` should be valid?

TangRufus replied:
Also syntax error.
Seems `...` must go last

swalkinshaw commented on cmd/valet_link.go:
```diff
@@ -0,0 +1,90 @@
+			valetLink = execCommand("valet", "link", "--secure", app)
+		}
+
+		valetLink.Dir = site.LocalPath
```
I think this is what's causing the testing issue but I'm not sure what the solution is yet

TangRufus replied:
Fixed by https://github.com/roots/trellis-cli/pull/48/commits/47e7137ddf2047c56d3da4cc27bbee3efea2efa8

swalkinshaw commented on cmd/valet_link.go:
```diff
@@ -0,0 +1,94 @@
+		return 1
+	}
+
+	c.UI.Info(fmt.Sprintf("Linking environment %s...\n", environment))
```
Based on your sample output, I might remove this newline and add one after "Linking site"?

TangRufus replied:
New output after https://github.com/roots/trellis-cli/pull/48/commits/c047d5b7a1dcbd8c8dc8e2f069df838dcfec5463:

```sh-session
$ trellis-cli valet link
Linking environment development...
Linking site no-enabled.com...

Running command => valet link no-enabled
A [no-enabled] symbolic link has been created in [/Users/me/.config/valet/Sites/no-enabled].
✓ Site no-enabled.com linked

Linking site insecure.com...

Running command => valet link insecure
A [insecure] symbolic link has been created in [/Users/me/.config/valet/Sites/insecure].
✓ Site insecure.com linked
```

Am I understanding it correctly?

swalkinshaw replied:
Yep looks great to me 👍

swalkinshaw commented on cmd/valet_link.go:
```diff
@@ -0,0 +1,94 @@
+		c.UI.Info(fmt.Sprintf("Linking site %s...", key))
+
+		canonical, _ := c.Trellis.HostsFromDomain(site.SiteHosts[0].Canonical, environment)
+		app := strings.TrimSuffix(canonical.String(), "." + canonical.TLD)
```
I *think* this might just be `canonical.SLD`? ie: `example` from `example.com`

TangRufus replied:
```yaml
wordpress_sites:
  third.second.com:
    site_hosts:
      - canonical: third.second.test
    local_path: ../bedrock-third-second
    ssl:
      enabled: true
```

Adding debug messages:
```goland
		canonical, _ := c.Trellis.HostsFromDomain(site.SiteHosts[0].Canonical, environment)
		app := strings.TrimSuffix(canonical.String(), "." + canonical.TLD)

		c.UI.Info(fmt.Sprintf("canonical:  %s", canonical))
		c.UI.Info(fmt.Sprintf("canonical.String():  %s", canonical.String()))

		c.UI.Info(fmt.Sprintf("canonical.TLD:  %s", canonical.TLD))
		c.UI.Info(fmt.Sprintf("canonical.SLD:  %s", canonical.SLD))
		c.UI.Info(fmt.Sprintf("canonical.TRD:  %s", canonical.TRD))

		c.UI.Info(fmt.Sprintf("app:  %s", app))
```

Output:
```
Linking site third.second.com...
canonical:  third.second.test
canonical.String():  third.second.test
canonical.TLD:  test
canonical.SLD:  second
canonical.TRD:  third
app:  third.second
```

The goal of `app := strings.TrimSuffix(canonical.String(), "." + canonical.TLD)` is to remove `.test` from `canonical` while keeping the subdomain part (TRD).

I think we should allow users to use subdomains because valet documented that:

> You can use valet link to serve the same project from multiple (sub)domains. To add a subdomain or another domain to your project run valet link subdomain.app-name from the project folder.
>
> -- https://laravel.com/docs/5.8/valet#serving-sites

swalkinshaw replied:
Ah I didn't think about subdomains. Whatever you think is best then. So what's where works

swalkinshaw commented on cmd/valet_link_test.go:
```diff
@@ -0,0 +1,189 @@
+
+import (
+	"os/exec"
+	//"os/exec"
```
🔥

TangRufus replied:
shame on me

swalkinshaw commented:
Awesome work thanks 🎉

TangRufus commented:
rebased and added https://github.com/roots/trellis-cli/pull/48/commits/86464df01cd87766c72aac4afaa3c28e7bf4f002 & https://github.com/roots/trellis-cli/pull/48/commits/a05afc3cc6e3be24192ab7e7fd1f60320c2c73ff
